File: Volta/Volta/Calculations/KPICalculator.swift
import Foundation

/// Pure Swift — no SwiftUI, no SwiftData. All inputs are plain Doubles.
enum KPICalculator {

    /// Bruttorendite = (Kaltmiete jährlich + Parkingmiete jährlich) / Kaufpreis
    static func grossYield(coldRentYearly: Double, parkingRentYearly: Double, purchasePrice: Double) -> Double? {
        guard purchasePrice > 0 else { return nil }
        return (coldRentYearly + parkingRentYearly) / purchasePrice
    }

    /// Nettorendite = NOI / Gesamtinvestment
    static func netYield(netOperatingIncomeYearly: Double, totalInvestment: Double) -> Double? {
        guard totalInvestment > 0 else { return nil }
        return netOperatingIncomeYearly / totalInvestment
    }

    /// Cap Rate = NOI / Kaufpreis (ohne Nebenkosten)
    static func capRate(netOperatingIncomeYearly: Double, purchasePrice: Double) -> Double? {
        guard purchasePrice > 0 else { return nil }
        return netOperatingIncomeYearly / purchasePrice
    }

    /// Cash-on-Cash Return = Cashflow nach Schuldendienst / eingesetztes EK
    static func cashOnCashReturn(cashflowAfterDebtYearly: Double, equityUsed: Double) -> Double? {
        guard equityUsed > 0 else { return nil }
        return cashflowAfterDebtYearly / equityUsed
    }

    /// DSCR (NOI-basiert) = NOI / jährlicher Schuldendienst
    static func dscrNOI(netOperatingIncomeYearly: Double, debtServiceAnnual: Double) -> Double? {
        guard debtServiceAnnual > 0 else { return nil }
        return netOperatingIncomeYearly / debtServiceAnnual
    }

    /// Mietmultiplikator = Kaufpreis / Jahreskaltmiete (inkl. Parking)
    static func mietmultiplikator(purchasePrice: Double, coldRentYearly: Double, parkingRentYearly: Double) -> Double? {
        let totalRent = coldRentYearly + parkingRentYearly
        guard totalRent > 0 else { return nil }
        return purchasePrice / totalRent
    }

    /// Break-Even-Miete = alle Eigentümer-Kosten im Vermietungsszenario + Kreditrate
    /// Enthält NICHT: propertyTaxAnnual (WE) und hoaFeeRecoverableMonthly (WE) — Mieter zahlt
    static func breakEvenRent(
        hoaFeeNonRecoverableMonthly: Double,
        hoaFeeMaintenanceReserveMonthly: Double,
        hoaFeeParkingNonRecoverableMonthly: Double,
        hoaFeeParkingRecoverableMonthly: Double,
        hoaFeeParkingMaintenanceReserveMonthly: Double,
        propertyTaxParkingAnnual: Double,
        propertyManagementAnnual: Double,
        propertyInsuranceAnnual: Double,
        otherCostsMonthly: Double,
        monthlyMortgage: Double,
        hasParking: Bool
    ) -> Double {
        var result = hoaFeeNonRecoverableMonthly
            + hoaFeeMaintenanceReserveMonthly
            + (propertyManagementAnnual / 12)
            + (propertyInsuranceAnnual / 12)
            + otherCostsMonthly
            + monthlyMortgage
        if hasParking {
            result += hoaFeeParkingNonRecoverableMonthly
                + hoaFeeParkingRecoverableMonthly
                + hoaFeeParkingMaintenanceReserveMonthly
                + (propertyTaxParkingAnnual / 12)
        }
        return result
    }

    /// Tatsächliche Leerstandsquote seit Besitzübergang bis heute
    /// Gibt nil zurück wenn keine StatusEntries vorhanden oder gesamtEigentumstage == 0
    static func actualVacancyRate(
        statusEntries: [StatusEntry],
        economicTransferDate: Date
    ) -> Double? {
        guard !statusEntries.isEmpty else { return nil }
        let today = Date()
        let calendar = Calendar.current
        let totalDays = calendar.dateComponents([.day], from: economicTransferDate, to: today).day ?? 0
        guard totalDays > 0 else { return nil }

        // Sort entries by date ascending
        let sorted = statusEntries.sorted { $0.date < $1.date }

        var vacancyDays = 0
        for (index, entry) in sorted.enumerated() {
            let segmentStart = max(entry.date, economicTransferDate)
            let segmentEnd: Date
            if index + 1 < sorted.count {
                segmentEnd = sorted[index + 1].date
            } else {
                segmentEnd = today
            }
            guard segmentEnd > segmentStart else { continue }
            if entry.status == .leerstand || entry.status == .mietgarantie {
                let days = calendar.dateComponents([.day], from: segmentStart, to: segmentEnd).day ?? 0
                vacancyDays += days
            }
        }
        return Double(vacancyDays) / Double(totalDays)
    }

    /// Wertsteigerung = aktueller Marktwert - Gesamtkaufpreis
    /// Gibt nil zurück wenn currentMarketValue nil, <= 0 oder totalPurchasePrice <= 0
    static func capitalGain(
        currentMarketValue: Double?,
        totalPurchasePrice: Double
    ) -> (absolute: Double, percent: Double)? {
        guard let currentMarketValue, currentMarketValue > 0, totalPurchasePrice > 0 else { return nil }
        let absolute = currentMarketValue - totalPurchasePrice
        let percent = absolute / totalPurchasePrice
        return (absolute, percent)
    }

    // MARK: - Intermediate helpers (used by ViewModels)

    /// Effektives Bruttoeinkommen = Bruttomiete * (1 - Leerstandsquote)
    static func effectiveGrossIncomeYearly(grossIncomeYearly: Double, vacancyRate: Double) -> Double {
        grossIncomeYearly * (1.0 - vacancyRate)
    }

    /// NOI = effektives Bruttoeinkommen - nicht-umlagefähige Kosten
    static func netOperatingIncomeYearly(effectiveGrossIncomeYearly: Double,
                                          operatingCostsNonRecoverableYearly: Double) -> Double {
        effectiveGrossIncomeYearly - operatingCostsNonRecoverableYearly
    }

    /// Cashflow nach Schuldendienst (jährlich)
    static func cashflowAfterDebtYearly(netOperatingIncomeYearly: Double, debtServiceAnnual: Double) -> Double {
        netOperatingIncomeYearly - debtServiceAnnual
    }

    /// Eingesetztes Eigenkapital = Gesamtinvestment - Darlehen
    static func equityUsed(totalInvestment: Double, loanAmount: Double) -> Double {
        totalInvestment - loanAmount
    }

    /// Nicht-umlagefähige Betriebskosten monatlich
    static func operatingCostsNonRecoverableMonthly(hoaFeeNonRecoverable: Double,
                                                      maintenanceReserve: Double,
                                                      propertyManagementMonthly: Double,
                                                      otherCostsMonthly: Double) -> Double {
        hoaFeeNonRecoverable + maintenanceReserve + propertyManagementMonthly + otherCostsMonthly
    }

    /// Umlagefähige Kosten monatlich (Mieter zahlt bei Vermietung)
    static func operatingCostsRecoverableMonthly(hoaFeeRecoverable: Double,
                                                   propertyTaxMonthly: Double,
                                                   propertyInsuranceMonthly: Double) -> Double {
        hoaFeeRecoverable + propertyTaxMonthly + propertyInsuranceMonthly
    }

    /// Gesamtinvestment = Kaufpreis + Kaufnebenkosten + Renovierung
    static func totalInvestment(purchasePrice: Double, closingCostsTotal: Double,
                                 renovationModernizationCosts: Double) -> Double {
        purchasePrice + closingCostsTotal + renovationModernizationCosts
    }

    /// Kaufnebenkosten gesamt
    static func closingCostsTotal(landTransferTax: Double, notaryCosts: Double,
                                   landRegistryCosts: Double, agentFee: Double,
                                   appraisalCosts: Double) -> Double {
        landTransferTax + notaryCosts + landRegistryCosts + agentFee + appraisalCosts
    }

    /// LTV = Restschuld / Gesamtinvestment
    static func ltvRatio(remainingDebt: Double, totalInvestment: Double) -> Double? {
        guard totalInvestment > 0 else { return nil }
        return remainingDebt / totalInvestment
    }
}

File: web/lib/calculations/kpiCalculator.ts
/** Bruttorendite = (Kaltmiete jährlich + Stellplatzmiete jährlich) / Kaufpreis */
export function grossYield(coldRentYearly: number, parkingRentYearly: number, purchasePrice: number): number | null {
  if (purchasePrice <= 0) return null;
  return (coldRentYearly + parkingRentYearly) / purchasePrice;
}

/** Nettorendite = NOI / Gesamtinvestment */
export function netYield(netOperatingIncomeYearly: number, totalInvestment: number): number | null {
  if (totalInvestment <= 0) return null;
  return netOperatingIncomeYearly / totalInvestment;
}

/** Cap Rate = NOI / Kaufpreis (ohne Nebenkosten) */
export function capRate(netOperatingIncomeYearly: number, purchasePrice: number): number | null {
  if (purchasePrice <= 0) return null;
  return netOperatingIncomeYearly / purchasePrice;
}

/** Cash-on-Cash Return = Cashflow nach Schuldendienst / eingesetztes EK */
export function cashOnCashReturn(cashflowAfterDebtYearly: number, equityUsed: number): number | null {
  if (equityUsed <= 0) return null;
  return cashflowAfterDebtYearly / equityUsed;
}

/** DSCR (NOI-basiert) = NOI / jährlicher Schuldendienst */
export function dscrNOI(netOperatingIncomeYearly: number, debtServiceAnnual: number): number | null {
  if (debtServiceAnnual <= 0) return null;
  return netOperatingIncomeYearly / debtServiceAnnual;
}

/** Mietmultiplikator = Kaufpreis / Jahreskaltmiete (inkl. Parking) */
export function mietmultiplikator(
  purchasePrice: number,
  coldRentYearly: number,
  parkingRentYearly: number
): number | null {
  const totalRent = coldRentYearly + parkingRentYearly;
  if (totalRent <= 0) return null;
  return purchasePrice / totalRent;
}

/** Break-Even-Miete = nicht-umlagefähige Kosten + Kreditrate */
export function breakEvenRentMonthly(operatingCostsNonRecoverableMonthly: number, monthlyMortgage: number): number {
  return operatingCostsNonRecoverableMonthly + monthlyMortgage;
}

/** LTV = Restschuld / Gesamtinvestment */
export function ltvRatio(remainingDebt: number, totalInvestment: number): number | null {
  if (totalInvestment <= 0) return null;
  return remainingDebt / totalInvestment;
}

/** Effektives Bruttoeinkommen = Bruttomiete * (1 - Leerstandsquote) */
export function effectiveGrossIncomeYearly(grossIncomeYearly: number, vacancyRate: number): number {
  return grossIncomeYearly * (1 - vacancyRate);
}

/** NOI = effektives Bruttoeinkommen - nicht-umlagefähige Kosten */
export function netOperatingIncomeYearly(
  effectiveGrossIncome: number,
  operatingCostsNonRecoverableYearly: number
): number {
  return effectiveGrossIncome - operatingCostsNonRecoverableYearly;
}

/** Eingesetztes Eigenkapital = Gesamtinvestment - Darlehen */
export function equityUsed(totalInvestment: number, loanAmount: number): number {
  return totalInvestment - loanAmount;
}

/** Gesamtinvestment = Kaufpreis + Kaufnebenkosten + Renovierung */
export function totalInvestment(
  purchasePrice: number,
  closingCostsTotal: number,
  renovationModernizationCosts: number
): number {
  return purchasePrice + closingCostsTotal + renovationModernizationCosts;
}

/** Hausgeld nicht umlagefähig = gesamt − umlagefähig − Instandhaltungsrücklage. */
export function hoaNonRecoverableMonthly(totalMonthly: number, recoverableMonthly: number, maintenanceReserveMonthly: number): number {
  return totalMonthly - recoverableMonthly - maintenanceReserveMonthly;
}

/** Kaufnebenkosten gesamt */
export function closingCostsTotal(
  landTransferTax: number,
  notaryCosts: number,
  landRegistryCosts: number,
  agentFee: number,
  appraisalCosts: number
): number {
  return landTransferTax + notaryCosts + landRegistryCosts + agentFee + appraisalCosts;
}

/** Tatsächliche Leerstandsquote = Leerstandstage / Eigentumstage seit Erwerb. */
export function actualVacancyRate(leerstandDays: number, ownershipDays: number): number | null {
  if (ownershipDays <= 0) return null;
  return leerstandDays / ownershipDays;
}

export type BenchmarkKpi =
  | 'grossYield'
  | 'netYield'
  | 'cashOnCash'
  | 'eigenkapitalrendite'
  | 'kaufpreisfaktor'
  | 'dscr'
  | 'ltv'
  | 'actualVacancyRate';
export type BenchmarkColor = 'green' | 'orange' | 'red';

interface BenchmarkThreshold {
  direction: 'higherIsBetter' | 'lowerIsBetter';
  green: number;
  orange: number;
  domainMin: number;
  domainMax: number;
}

// Per spec-overview-tab.md's 3-tier chip table (grün/orange/rot) — the richer
// 4-tier "Kontext" copy in docs/superpowers/specs/2026-06-14-kpi-benchmarks.md
// feeds the KPI info sheet's text, not this coloring. domainMin/domainMax define
// the 0–100% range the KpiScale marker is placed within (see scalePosition below) —
// chosen with headroom beyond the orange threshold so realistic values don't sit
// permanently pinned to an edge; see docs/superpowers/specs/2026-08-01-uebersicht-redesign-design.md.
const BENCHMARK_THRESHOLDS: Record<BenchmarkKpi, BenchmarkThreshold> = {
  grossYield: { direction: 'higherIsBetter', green: 0.05, orange: 0.03, domainMin: 0, domainMax: 0.1 },
  netYield: { direction: 'higherIsBetter', green: 0.04, orange: 0.02, domainMin: 0, domainMax: 0.08 },
  cashOnCash: { direction: 'higherIsBetter', green: 0.06, orange: 0.03, domainMin: -0.2, domainMax: 0.2 },
  // Higher than cashOnCash's thresholds since this also credits Tilgung + Wertsteigerung.
  eigenkapitalrendite: { direction: 'higherIsBetter', green: 0.08, orange: 0.04, domainMin: -0.1, domainMax: 0.2 },
  kaufpreisfaktor: { direction: 'lowerIsBetter', green: 20, orange: 25, domainMin: 10, domainMax: 35 },
  dscr: { direction: 'higherIsBetter', green: 1.25, orange: 1.0, domainMin: 0, domainMax: 2.0 },
  ltv: { direction: 'lowerIsBetter', green: 0.7, orange: 0.8, domainMin: 0, domainMax: 1.1 },
  actualVacancyRate: { direction: 'lowerIsBetter', green: 0.03, orange: 0.08, domainMin: 0, domainMax: 0.2 },
};

/** Chip color for a KPI value. null value (no data yet) -> null (no chip rendered). */
export function benchmarkColor(kpi: BenchmarkKpi, value: number | null): BenchmarkColor | null {
  if (value === null) return null;
  const t = BENCHMARK_THRESHOLDS[kpi];
  if (t.direction === 'higherIsBetter') {
    if (value >= t.green) return 'green';
    if (value >= t.orange) return 'orange';
    return 'red';
  }
  if (value <= t.green) return 'green';
  if (value <= t.orange) return 'orange';
  return 'red';
}

/** Read-only access to a KPI's threshold/domain config, e.g. for rendering scale axis labels. */
export function benchmarkThreshold(kpi: BenchmarkKpi): Readonly<BenchmarkThreshold> {
  return BENCHMARK_THRESHOLDS[kpi];
}

/**
 * Where `value` sits on a 0–1 scale within [domainMin, domainMax], oriented so
 * "good" is always 1 (right/green end) and "bad" is always 0 (left/red end) —
 * regardless of the KPI's direction. Out-of-domain values clamp to the nearest end.
 */
export function scalePosition(kpi: BenchmarkKpi, value: number): number {
  const t = BENCHMARK_THRESHOLDS[kpi];
  const raw =
    t.direction === 'higherIsBetter'
      ? (value - t.domainMin) / (t.domainMax - t.domainMin)
      : (t.domainMax - value) / (t.domainMax - t.domainMin);
  return Math.min(1, Math.max(0, raw));
}
